src/cuda/kernels/complex.f90

```diff
@@ -0,0 +1,295 @@
+module m_cuda_complex
+   use cudafor
+
+   use m_common, only: dp
+   use m_cuda_common, only: SZ
+
+   implicit none
+
+contains
+
+   attributes(global) subroutine processfftdiv( &
+         div, waves, nx, ny, nz, ax, bx, ay, by, az, bz &
+         )
+      implicit none
+
+      ! Arguments
+      complex(dp), device, intent(inout), dimension(:, :, :) :: div
+      complex(dp), device, intent(in), dimension(:, :, :) :: waves
+      real(dp), device, intent(in), dimension(:) :: ax, bx, ay, by, az, bz
+      integer, value, intent(in) :: nx, ny, nz
+
+      ! Local variables
+      integer :: i, j, b, ix, iy, iz
+      real(dp) :: tmp_r, tmp_c, div_r, div_c
+
+      i = threadIdx%x
+      b = blockIdx%x
+
+      do j = 1, nx
+         ! normalisation
+         div_r = real(div(i, j, b), kind=dp)/(nx*ny*nz)
+         div_c = aimag(div(i, j, b))/(nx*ny*nz)
+
+         ! get the indices for x, y, z directions
+         ix = j; iy = i + (b - 1)/(nz/2 + 1)*SZ; iz = mod(b - 1, nz/2 + 1) + 1
+
+         ! post-process forward
+         ! post-process in z
+         tmp_r = div_r
+         tmp_c = div_c
+         div_r = tmp_r*bz(iz) + tmp_c*az(iz)
+         div_c = tmp_c*bz(iz) - tmp_r*az(iz)
+
+         ! post-process in y
+         tmp_r = div_r
+         tmp_c = div_c
+         div_r = tmp_r*by(iy) + tmp_c*ay(iy)
+         div_c = tmp_c*by(iy) - tmp_r*ay(iy)
+         if (iy > ny/2 + 1) div_r = -div_r
+         if (iy > ny/2 + 1) div_c = -div_c
+
+         ! post-process in x
+         tmp_r = div_r
+         tmp_c = div_c
+         div_r = tmp_r*bx(ix) + tmp_c*ax(ix)
+         div_c = tmp_c*bx(ix) - tmp_r*ax(ix)
+         if (ix > nx/2 + 1) div_r = -div_r
+         if (ix > nx/2 + 1) div_c = -div_c
+
+         ! Solve Poisson
+         tmp_r = real(waves(i, j, b), kind=dp)
+         tmp_c = aimag(waves(i, j, b))
+         if ((tmp_r < 1.e-16_dp) .or. (tmp_c < 1.e-16_dp)) then
+            div_r = 0._dp; div_c = 0._dp
+         else
+            div_r =-div_r/tmp_r
+            div_c =-div_c/tmp_c
+         end if
+
+         ! post-process backward
+         ! post-process in z
+         tmp_r = div_r
+         tmp_c = div_c
+         div_r = tmp_r*bz(iz) - tmp_c*az(iz)
+         div_c =-tmp_c*bz(iz) - tmp_r*az(iz)
+
+         ! post-process in y
+         tmp_r = div_r
+         tmp_c = div_c
+         div_r = tmp_r*by(iy) + tmp_c*ay(iy)
+         div_c = tmp_c*by(iy) - tmp_r*ay(iy)
+         if (iy > ny/2 + 1) div_r = -div_r
+         if (iy > ny/2 + 1) div_c = -div_c
+
+         ! post-process in x
+         tmp_r = div_r
+         tmp_c = div_c
+         div_r = tmp_r*bx(ix) + tmp_c*ax(ix)
+         div_c =-tmp_c*bx(ix) + tmp_r*ax(ix)
+         if (ix > nx/2 + 1) div_r = -div_r
+         if (ix > nx/2 + 1) div_c = -div_c
+
+         ! update the entry
+         div(i, j, b) = cmplx(div_r, div_c, kind=dp)
+      end do
+
+   end subroutine processfftdiv
+
+   attributes(global) subroutine reorder_cmplx_x2y_T(u_y, u_x, nz)
```

Comment:
Would there be a way to not rewrite entirely the reorder functions just because we are dealing with complex here?
Also, It would be good to test those as well. Once again, they are very easy to get slightly wrong.

Reply:
Testing functions we're likely to throw away seems a waste.. but throwing away the functions also seems potentially wasteful :shrug: 

@semi-h Why do the reorder fns need to be thrown away when cuFFTMP comes in?

Reply:
We just need to transform the divergence of velocity into spectral space (3D forward FFT) so that we can postprocess it to obtain pressure field in spectral space. Then we apply an inverse FFT in 3D and bring the pressure field from spectral space back to physical space. Currently we implement our own 3D FFT transforms, both forward and inverse. And this requires a series of 1D FFT calls and those reshapes/reorders in between.

cuFFTMp will allow us to carry out the 3D forward or inverse FFTs with a single call to the library.